New messaging specification (replacing old deposit/withdrawal spec) (#318)
FuelLabs/fuel-specs: specs/protocol/tx_format.md

- [Input](#input)
  - [InputCoin](#inputcoin)
  - [InputContract](#inputcontract)
  - [InputMessage](#inputmessage)
- [OutputType](#outputtype)
- [Output](#output)
  - [OutputCoin](#outputcoin)
  - [OutputContract](#outputcontract)

>   - [OutputMessage](#outputmessage)

Reviewer: Hmmm. Should we have an output message and an output message variable? Would be nice for users to potentially add messages to their tx without a script. Not strictly necessary.

Author: We could. At the very least maybe we make the `SMO` opcode available to transaction scripts. I also feel like messages output from smart contracts should be the default so we should just call "OutputMessageVariable" `OutputMessage` and if we make an option for user to sign messages without a script we call it "OutputMessageDirect" or "OutputMessageImmediate" or something

Reviewer: We can always add the direct message in a future PR if it's needed, since it's conceptually a simple change. Let's keep this one simple.

| name   | type                                                              | description    |
|--------|-------------------------------------------------------------------|----------------|
| `type` | [InputType](#inputtype)                                           | Type of input. |

> | `data` | One of [InputCoin](#inputcoin), [InputContract](#inputcontract) or [InputMessage](#inputmessage) | Input data.    |

Reviewer: ```suggestion
| `data` | One of [InputCoin](#inputcoin), [InputContract](#inputcontract), or [InputMessage](#inputmessage) | Input data.    |
```

Note: when executing a script, `txID`, `outputIndex`, `balanceRoot`, and `stateRoot` are initialized to the transaction ID, output index, amount, and state root of the contract with ID `contractID`, and `txPointer` is initialized to zero.

### InputMessage

| name                  | type        | description                                                            |
|-----------------------|-------------|------------------------------------------------------------------------|
| `messageID`           | `byte[32]`  | The messageID as described [here](./identifiers.md#message-id).        |
| `sender`              | `byte[32]`  | The address of the message sender.                                     |
| `recipient`           | `byte[32]`  | The address of the message recipient.                                  |

> | `dataLength`          | `uint16`    | Length of message data, in bytes.                                      |

Reviewer: I would put this right before `predicateLength`.

| `amount`              | `uint64`    | Amount of base asset coins sent with message.                          |
| `nonce`               | `byte[32]`  | The message nonce.                                                     |
| `owner`               | `byte[32]`  | Owning address or predicate hash.                                      |
| `witnessIndex`        | `uint8`     | Index of witness that authorizes spending the coin.                    |
| `predicateLength`     | `uint16`    | Length of predicate, in instructions.                                  |
| `predicateDataLength` | `uint16`    | Length of predicate input data, in bytes.                              |
| `data`                | `byte[]`    | The message data.                                                      |
| `predicate`           | `byte[]`    | Predicate bytecode.                                                    |
| `predicateData`       | `byte[]`    | Predicate input data (parameters).                                     |

> Transaction is invalid if:

Reviewer: Hmmm. There should be a bound here on the data length. Maybe make a new consensus parameter, `MAX_MESSAGE_DATA_LENGTH`.

Note: when executing a script, `txID`, `outputIndex`, `balanceRoot`, and `stateRoot` are initialized to the transaction ID, output index, amount, and state root of the contract with ID `contractID`, and `txPointer` is initialized to zero.

### InputMessage

| name                  | type        | description                                                            |
|-----------------------|-------------|------------------------------------------------------------------------|
| `messageID`           | `byte[32]`  | The messageID as described [here](./identifiers.md#message-id).        |
| `sender`              | `byte[32]`  | The address of the message sender.                                     |
| `recipient`           | `byte[32]`  | The address of the message recipient.                                  |
| `dataLength`          | `uint16`    | Length of message data, in bytes.                                      |

> | `data`                | `byte[]`    | The message data or [abi encoded](./abi.md) call to execute.           |

Reviewer: We don't really need Solidity ABI encoding here I think? It would be Fuel encoded data, but that's application logic. The protocol rules don't care about what the data is.
```suggestion
| `data`                | `byte[]`    | The message data .           |
```

Author: It's referring to Fuels abi with the thought being that most messages will be going through a standard relayer contract where the message data will be the abi for calling a Fuel contract function. But I agree that this is application specific and shouldn't be in the spec.

> | `data`                | `byte[]`    | The message data or [abi encoded](./abi.md) call to execute.           |

Reviewer: Move this to the back. Note that fixed-length field are first, then all variable-length fields. This is required to be able to read fixed-length fields in constant time.

| `amount`              | `uint64`    | Amount of base asset coins sent with message.                          |
| `nonce`               | `uint64`    | The message nonce.                                                     |
| `owner`               | `byte[32]`  | Owning address or predicate hash.                                      |
| `witnessIndex`        | `uint8`     | Index of witness that authorizes spending the coin.                    |
| `predicateLength`     | `uint16`    | Length of predicate, in instructions.                                  |
| `predicateDataLength` | `uint16`    | Length of predicate input data, in bytes.                              |
| `predicate`           | `byte[]`    | Predicate bytecode.                                                    |
| `predicateData`       | `byte[]`    | Predicate input data (parameters).                                     |

Transaction is invalid if:

- `witnessIndex >= tx.witnessesCount`
- `predicateLength > MAX_PREDICATE_LENGTH`
- `predicateDataLength > MAX_PREDICATE_DATA_LENGTH`
- If `predicateLength > 0`; the computed predicate root (see below) is not equal `owner`

> Note: the predicate root is computed identically to the contract ID, [here](./identifiers.md#contract-id).

Reviewer: #321
```suggestion
The predicate root is computed identically to the contract root, used to compute the contract ID, [here](./identifiers.md#contract-id).
```

The state root `stateRoot` is the root of the [SMT](./cryptographic_primitives.md#sparse-merkle-tree) of storage slots. Each storage slot is a `byte[32]`, keyed by a `byte[32]`.

### OutputMessage

| name                  | type        | description                                                             |
|-----------------------|-------------|-------------------------------------------------------------------------|

> | `messageID`           | `byte[32]`  | The messageID as described [here](./identifiers.md#message-id).         |

Reviewer: Wait, why does this have a message ID? Isn't that only for inputs?

Author: Right now, the message ID for outputs is what's placed in the binary merkle tree that gets committed to Ethereum. It's these message IDs that the Inbox uses to ensure a message is never relayed more than once

Reviewer: Ah I see. In that case, is there really a need for the rest of the fields? Those can be placed in the receipt, rather than in the output.

Author: I thought we weren't doing receipts for messages because all the data is already in the output? The fields needs to be somewhere that's easy for a user to pull from a client so anyone can then relay the messages on L1

Author: I've now moved most of the OutputMessage data into a receipt. Data length is capped at 2**16 or 64kB
